### PR TITLE
[Airflow 16364] Add conn_timeout and cmd_timeout params to SSHOperator; add conn_timeout param to SSHHook

### LINKED ISSUE #16364: Timeout is ambiguous in SSHHook and SSHOperator
In SSHHook the timeout argument of the constructor is used to set a connection timeout. This is fine.

But in SSHOperator the timeout argument of the constructor is used for *both* the timeout of the SSHHook *and* the timeout of the command itself (see paramiko's ssh client exec_command use of the timeout parameter). This ambiguous use of the same parameter is very dirty.

I see two ways to clean the behaviour: 

1. Let the SSHHook constructor be the only way to handle the connection timeout (thus, if one wants a specific timeout they should explicitely build a hook to be passed to the operator using the operator's constructor).
2. Split the timeout argument in SSHOperator into two arguments conn_timeout and cmd_timeout for example.

The choice between 1 and 2 depends on how frequently people are supposed to want to change the connection timeout. If it is something very frequent. then go for 2. if not go for 1.

BR and thanks for the code!

### PR DESCRIPTION
**Pull request by fatmumuhomer**
closes #16364 (Timeout is ambiguous in SSHHook and SSHOperator)

Add conn_timeout and cmd_timeout params to SSHOperator. Add conn_timeout param to SSHHook. The legacy timeout param remains but is marked as deprecated. The new params take precedence over the legacy timeout one if both/all are provided.

---
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

### CODE REVIEW

File: airflow/providers/ssh/hooks/ssh.py [hunk 1]
```diff
@@ -56,8 +56,10 @@ class SSHHook(BaseHook):
     :type key_file: str
     :param port: port of remote host to connect (Default is paramiko SSH_PORT)
     :type port: int
-    :param timeout: timeout for the attempt to connect to the remote_host.
+    :param timeout: (Deprecated). timeout for the attempt to connect to the remote_host.
     :type timeout: int
+    :param conn_timeout: timeout for the attempt to connect to the remote_host.
+    :type conn_timeout: int
```
Comment (uranusjr) [inline airflow/providers/ssh/hooks/ssh.py]: We should put this above the deprecated `timeout`

File: airflow/providers/ssh/hooks/ssh.py [hunk 2]
```diff
@@ -101,7 +103,8 @@ def __init__(
         password: Optional[str] = None,
         key_file: Optional[str] = None,
         port: Optional[int] = None,
-        timeout: int = 10,
+        timeout: Optional[int] = 10,
```
Comment (uranusjr) [inline airflow/providers/ssh/hooks/ssh.py]: Is this change needed/correct? I feel we want to either change the default to `None` (so we can better detect whether `timeout` is passed in by the user), or the type should continue to be `int` (since the user shouldn’t do `timeout=None` in any situation).

File: airflow/providers/ssh/hooks/ssh.py [hunk 3]
```diff
@@ -185,6 +200,18 @@ def __init__(
                     self.host_key = key_constructor(data=decoded_host_key)
                     self.no_host_key_check = False
 
+        if self.timeout:
```
Comment (uranusjr) [inline airflow/providers/ssh/hooks/ssh.py:208]: Continuing above, this check seems to make more sense if the default `timeout` is `None`.

**Comment by uranusjr** [comment]
See inline comments (the same also apply to the operator class).

File: airflow/providers/ssh/hooks/ssh.py (hunk 2 above)
Comment (fatmumuhomer) [reply airflow/providers/ssh/hooks/ssh.py]: Yeah, looking at it again you're right. It should default to None otherwise the user will need to purposefully pass timeout = None to turn off the deprecation warning.

In order to maintain backwards compatibility, since the current versions of the hook/operator allow the user to omit timeout, I'll need to handle a situation where no timeout is passed and set the default later in the code.

**Comment by fatmumuhomer** [comment]
@uranusjr I made your suggested changes. Thanks for reviewing.

File: airflow/providers/ssh/hooks/ssh.py [hunk 4]
```diff
@@ -148,7 +155,17 @@ def __init__(
...
                 if "timeout" in extra_options:
-                    self.timeout = int(extra_options["timeout"], 10)
+                    warnings.warn(
+                        'Extra option `timeout` is deprecated.'
+                        'Please use `conn_timeout` instead.'
+                        'The old option `timeout` will be removed in a future version.',
+                        DeprecationWarning,
+                        stacklevel=2,
+                    )
+                    self.timeout = int(extra_options.get('timeout', 10))
```
Comment (uranusjr) [inline airflow/providers/ssh/hooks/ssh.py]:
```suggestion
                    self.timeout = int(extra_options['timeout'])
```

We already know `"timeout"` is in `extra_options` (the `if` check above), this `get()` is not needed.

File: airflow/providers/ssh/hooks/ssh.py [hunk 5]
```diff
@@ -148,7 +155,17 @@ def __init__(
...
+
+                if "conn_timeout" in extra_options:
+                    self.conn_timeout = int(extra_options.get('conn_timeout'))
```
Comment (uranusjr) [inline airflow/providers/ssh/hooks/ssh.py]:
```suggestion
                    self.conn_timeout = int(extra_options['conn_timeout'])
```

**Review by uranusjr — CHANGES_REQUESTED** [review]
Almost there!

**Comment by fatmumuhomer** [comment]
Thanks, @uranusjr! Those two changes are done.

**Review by uranusjr — APPROVED** [review]

**Comment by fatmumuhomer** [comment]
@uranusjr Thanks. What happens next? How do these changes get merged into the main project?

File: airflow/providers/ssh/hooks/ssh.py [hunk 6]
```diff
@@ -56,7 +58,9 @@ class SSHHook(BaseHook):
     :type key_file: str
     :param port: port of remote host to connect (Default is paramiko SSH_PORT)
     :type port: int
-    :param timeout: timeout for the attempt to connect to the remote_host.
+    :param conn_timeout: timeout for the attempt to connect to the remote_host.
+    :type conn_timeout: int
+    :param timeout: (Deprecated). timeout for the attempt to connect to the remote_host.
```
Comment (jedcunningham) [inline airflow/providers/ssh/hooks/ssh.py:65]: Should we mention what param to use instead?

File: airflow/providers/ssh/operators/ssh.py [hunk 7]
```diff
@@ -43,7 +46,11 @@ class SSHOperator(BaseOperator):
     :type remote_host: str
     :param command: command to execute on remote host. (templated)
     :type command: str
-    :param timeout: timeout (in seconds) for executing the command. The default is 10 seconds.
+    :param conn_timeout: timeout (in seconds) for maintaining the connection. The default is 10 seconds.
+    :type conn_timeout: int
+    :param cmd_timeout: timeout (in seconds) for executing the command. The default is 10 seconds.
+    :type cmd_timeout: int
+    :param timeout: (deprecated) timeout (in seconds) for executing the command. The default is 10 seconds.
```
Comment (jedcunningham) [inline airflow/providers/ssh/operators/ssh.py:55]: Here as well?

File: tests/providers/ssh/hooks/test_ssh.py [hunk 8]
```diff
@@ -475,6 +504,120 @@ def test_ssh_connection_with_no_host_key_where_no_host_key_check_is_false(self,
...
+            port='port',
+            username='username',
+            timeout=10,
+        )
+
+        with hook.get_conn():
+            ssh_mock.return_value.connect.assert_called_once_with(
+                hostname='remote_host',
+                username='username',
+                timeout=20,
```
Comment (jedcunningham) [inline tests/providers/ssh/hooks/test_ssh.py:581]: I realize the existing behavior was for extras to win, but that seems backwards to me. I'd expect the param to win.

Maybe we leave `timeout` as-is, but invert it for `conn_timeout` since it is new?

(`hostname` is a good example)

File: airflow/providers/ssh/hooks/ssh.py (hunk 6 above)
Comment (potiuk) [reply airflow/providers/ssh/hooks/ssh.py:65]: Yeah I think it is a bit ambiguous now. with both params in the docs :)

**Comment by potiuk** [comment]
Just add the comment in docs please and rebase and it should be cool to merge

**Review by potiuk — APPROVED** [review]

File: tests/providers/ssh/hooks/test_ssh.py (hunk 8 above)
Comment (fatmumuhomer) [reply tests/providers/ssh/hooks/test_ssh.py:581]: I'm in favor of that behavior change. I agree - it would make sense and be more transparent if the explicit parameters won. I'm happy to make the change. Any objections, @potiuk?

**Comment by potiuk** [comment]
Yep. I think it makes sense

**Comment by potiuk** [comment]
Some static checks are failing

**Comment by fatmumuhomer** [comment]
I'll take a look at the checks.

**Comment by fatmumuhomer** [comment]
I fixed the line that was failing flake8. Sorry about that.

**Comment by potiuk** [comment]
no problem . I really recommend installing pre-commit, it helps a lot in those cases even before you commit your changes :)

**Comment by fatmumuhomer** [comment]
Thanks - I'm going to do that :) Also, I just pushed the suggested by @jedcunningham to favor the conn_timeout param over the extras ready.

**Review by jedcunningham — CHANGES_REQUESTED** [review]
Overall LGTM. One more thing I just thought of - we need to add `conn_timeout` and mark `timeout` as deprecated in the docs for the connection:
https://github.com/apache/airflow/blob/main/docs/apache-airflow-providers-ssh/connections/ssh.rst

Might not be a bad idea to also specify whether it takes precedence over the hook param too.

**Comment by fatmumuhomer** [comment]
FYI - I added conn_timeout to ssh.rst alongside timeout. I considered using the deprecated directive, but it requires a version which hasn't been set yet. It wouldn't be a bad idea to update the rst again with the specific version once this change is officially released.

File: tests/providers/ssh/hooks/test_ssh.py [hunk 9]
```diff
@@ -475,6 +504,122 @@ def test_ssh_connection_with_no_host_key_where_no_host_key_check_is_false(self,
...
+        with hook.get_conn():
+            ssh_mock.return_value.connect.assert_called_once_with(
+                hostname='remote_host',
+                username='username',
+                timeout=15,
+                compress=True,
+                port='port',
+                sock=None,
+                look_for_keys=True,
+            )
```
Comment (jedcunningham) [inline tests/providers/ssh/hooks/test_ssh.py:632]: Aren't we missing a test where neither `timeout` and `conn_timeout` are passed and we use the value from the connection extras? Also missing a test where the default is used if the connection doesn't have one set?

File: airflow/providers/ssh/operators/ssh.py [hunk 10]
```diff
@@ -77,9 +87,24 @@ def __init__(
         self.remote_host = remote_host
         self.command = command
         self.timeout = timeout
+        self.conn_timeout = conn_timeout
+        self.cmd_timeout = cmd_timeout
+        if self.conn_timeout is None:
+            self.conn_timeout = self.timeout if self.timeout else TIMEOUT_DEFAULT
```
Comment (jedcunningham) [inline airflow/providers/ssh/operators/ssh.py]: I think by setting a default here, we effectively break allowing the connections `conn_timeout` to be used over the default. Maybe this would be better? Btw, I think we need better test coverage on default / extra handling like the hook too.

```suggestion
        if self.conn_timeout is None and self.timeout:
            self.conn_timeout = self.timeout
```

This'll let `None` be passed into the hook, which will let the connection extras be used and we avoid having the "default" set in 2 places.

File: airflow/providers/ssh/operators/ssh.py [hunk 11]
```diff
@@ -43,7 +46,12 @@ class SSHOperator(BaseOperator):
     :type remote_host: str
     :param command: command to execute on remote host. (templated)
     :type command: str
-    :param timeout: timeout (in seconds) for executing the command. The default is 10 seconds.
+    :param conn_timeout: timeout (in seconds) for maintaining the connection. The default is 10 seconds.
```
Comment (jedcunningham) [inline airflow/providers/ssh/operators/ssh.py:49]: Maybe the same language here as is used on the `remote_host`? "If provided, it will replace..."

File: docs/apache-airflow-providers-ssh/connections/ssh.rst [hunk 12]
```diff
@@ -47,7 +47,8 @@ Extra (optional)
     * ``key_file`` - Full Path of the private SSH Key file that will be used to connect to the remote_host.
     * ``private_key`` - Content of the private key used to connect to the remote_host.
     * ``private_key_passphrase`` - Content of the private key passphrase used to decrypt the private key.
-    * ``timeout`` - An optional timeout (in seconds) for the TCP connect. Default is ``10``.
+    * ``conn_timeout`` - An optional timeout (in seconds) for the TCP connect. Default is ``10``. The paramater to :class:`~airflow.providers.ssh.hooks.ssh.SSHHook` conn_timeout takes precedence.
```
Comment (jedcunningham) [inline docs/apache-airflow-providers-ssh/connections/ssh.rst]: Maybe we should document on the other side instead of here, like `remote_host`? Thoughts?

**Review by jedcunningham — CHANGES_REQUESTED** [review]
Spotted a couple more things @fatmumuhomer. I think we are getting close now 👍

File: airflow/providers/ssh/operators/ssh.py (hunk 10 above)
Comment (fatmumuhomer) [reply airflow/providers/ssh/operators/ssh.py]: Good catch! A test around this case would have caught this, too, so my fault for not creating one. I'll tackle both of those tonight.

File: docs/apache-airflow-providers-ssh/connections/ssh.rst (hunk 12 above)
Comment (fatmumuhomer) [reply docs/apache-airflow-providers-ssh/connections/ssh.rst]: Sorry - what do you mean by "on the other side"?

File: tests/providers/ssh/hooks/test_ssh.py (hunk 9 above)
Comment (fatmumuhomer) [reply tests/providers/ssh/hooks/test_ssh.py:632]: I was thinking about this a bit more - within SSHHook there are 16 ways to combine the two parameters and the two extra options. To be safe, the tests should really check them all to ensure we're selecting the value we expect each time. I'm working on a parametrized version that covers all 16 combinations.

Comment (jedcunningham) [reply tests/providers/ssh/hooks/test_ssh.py:632]: Cool, sounds good 👍

File: docs/apache-airflow-providers-ssh/connections/ssh.rst (hunk 12 above)
Comment (jedcunningham) [reply docs/apache-airflow-providers-ssh/connections/ssh.rst]: My bad, I meant on the hook/operator side like was suggested for the operator in my other comment.

File: tests/providers/ssh/hooks/test_ssh.py (hunk 9 above)
Comment (fatmumuhomer) [reply tests/providers/ssh/hooks/test_ssh.py:632]: The parametrized tests are done and I removed the non-parametrized ones that were now duplicates.

**Comment by fatmumuhomer** [comment]
@jedcunningham I added the parametrized SSHHook tests and also adjusted the rst and moved the details about conn_timeout to SSHHook as suggested. Sorry for the delay - August was a little crazy at work.

**Comment by uranusjr** [comment]
Any further thoughts @jedcunningham?

File: tests/providers/ssh/operators/test_ssh.py [hunk 13]
```diff
@@ -138,7 +160,9 @@ def test_no_output_command(self):
     def test_arg_checking(self):
```
Comment (jedcunningham) [inline tests/providers/ssh/operators/test_ssh.py]: I'm wondering if we should split this apart? We now have 7 different cases in a single test.

**Review by jedcunningham — APPROVED** [review]
There is one test that I think might be worth splitting up and there is a conflict now, otherwise LGTM.

File: tests/providers/ssh/operators/test_ssh.py (hunk 13 above)
Comment (fatmumuhomer) [reply tests/providers/ssh/operators/test_ssh.py]: If I understand what github is telling me, it looks like I can't be the one to resolve the conflict.

If I have some time tonight I'll split this test apart. I'm also happy to open a new PR for it if we want to close this one - whatever is easier for you all.

Comment (jedcunningham) [reply tests/providers/ssh/operators/test_ssh.py]: For the conflict, you should be able to resolve it by rebasing on the current main branch.

Comment (fatmumuhomer) [reply tests/providers/ssh/operators/test_ssh.py]: Alright, I **think** it's good now. I rebased and merged my changes to the file back in and there are no more conflicts. I hope I followed the right process - if not please let me know!

**Comment by potiuk** [comment]
Woohooo!

**Comment by ashb** [comment]
Also congrats on being the contributor who edged Airflow ahead of Spark in terms of number of contributors :D 

image

image

**Comment by fatmumuhomer** [comment]
@ashb That's a pretty cool honor :)

Thanks to everyone who helped me with this!

**Comment by potiuk** [comment]
How cool's that !
